pkg/object/webdav.go

```diff
@@ -85,103 +111,162 @@ func (w *webdav) Get(key string, off, limit int64) (io.ReadCloser, error) {
 	if err != nil {
 		return nil, err
 	}
-	if resp.StatusCode != 200 && resp.StatusCode != 206 {
-		return nil, parseError(resp)
+	if resp.StatusCode == http.StatusPartialContent {
+		// server supported partial content, return as-is.
+		return resp.Body, nil
 	}
-	return resp.Body, nil
-}
 
-func (w *webdav) mkdirs(p string) error {
-	err := w.c.Mkdir(p)
-	if err != nil && w.isNotExist(path.Dir(p)) {
-		if w.mkdirs(path.Dir(p)) == nil {
-			err = w.c.Mkdir(p)
+	// server returned success, but did not support partial content, so we have the whole
+	// stream in rs.Body
+	if resp.StatusCode == 200 {
+		if _, err := io.Copy(io.Discard, io.LimitReader(resp.Body, off)); err != nil {
+			return nil, &os.PathError{Op: "ReadStreamRange", Path: key, Err: err}
 		}
+		// return a io.ReadCloser that is limited to `length` bytes.
+		return &limitedReadCloser{resp.Body, int(limit)}, nil
 	}
-	return err
+	_ = resp.Body.Close()
+	return nil, &os.PathError{Op: "ReadStreamRange", Path: key, Err: err}
 }
 
-func (w *webdav) isNotExist(key string) bool {
-	if _, err := w.c.Stat(key); err != nil {
-		return strings.Contains(strings.ToLower(err.Error()), "not found")
+func (w *webdav) Put(key string, in io.Reader) error {
+	if key == "" {
+		return nil
+	}
+	if strings.HasSuffix(key, dirSuffix) {
+		return w.c.MkdirAll(key, 0)
 	}
-	return false
+	return w.c.WriteStream(key, in, 0)
 }
 
-func (w *webdav) path(key string) string {
-	return "/" + key
+func (w *webdav) Delete(key string) error {
+	if strings.HasSuffix(key, dirSuffix) {
```

Comment:
put  `a/b/c/d`   and then delete `a/b/c` , `a/b/c`  is a directory with files, but will be deleted.